From ff54e51d67b213aef50b142c967b8aa367da5a57 Mon Sep 17 00:00:00 2001
From: David Weinehall <david.weinehall@linux.intel.com>
Date: Thu, 13 Aug 2015 16:14:15 +0300
Subject: [PATCH] drm/i915: Allow parsing of variable size child device entries
 from VBT, addendum v2

Expand common_child_dev_config to be able to fit all information
defined by the latest VBT specification.

v2: Use proper conversion specifier for size_t (%zu)

Signed-off-by: David Weinehall <david.weinehall@linux.intel.com>
CC: Antti Koskipaa <antti.koskipaa@linux.intel.com>

 intel_bios.c |    7 ++++++-
 intel_bios.h |    4 ++++
 2 files changed, 10 insertions(+), 1 deletion(-)
---
 drivers/gpu/drm/i915/intel_bios.c | 7 ++++++-
 drivers/gpu/drm/i915/intel_bios.h | 4 ++++
 2 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/i915/intel_bios.c b/drivers/gpu/drm/i915/intel_bios.c
index 3dcd59e..2cbdd95 100644
--- a/drivers/gpu/drm/i915/intel_bios.c
+++ b/drivers/gpu/drm/i915/intel_bios.c
@@ -1083,6 +1083,10 @@ parse_device_mapping(struct drm_i915_private *dev_priv,
 		DRM_DEBUG_KMS("No general definition block is found, no devices defined.\n");
 		return;
 	}
+	/* Remember to keep this in sync with child_device_config;
+	 * whenever a new feature is added to BDB that causes that
+	 * struct to grow this needs to be updated too
+	 */
 	if (bdb->version < 195) {
 		expected_size = 33;
 	} else if (bdb->version == 195) {
@@ -1096,7 +1100,8 @@ parse_device_mapping(struct drm_i915_private *dev_priv,
 	}
 
 	if (expected_size > sizeof(*p_child)) {
-		DRM_ERROR("child_device_config cannot fit in p_child\n");
+		DRM_ERROR("child_device_config (size %u) cannot fit in p_child (size %zu); bdb->version: %u\n",
+			  expected_size, sizeof(*p_child), bdb->version);
 		return;
 	}
 
diff --git a/drivers/gpu/drm/i915/intel_bios.h b/drivers/gpu/drm/i915/intel_bios.h
index af0b476..4211737 100644
--- a/drivers/gpu/drm/i915/intel_bios.h
+++ b/drivers/gpu/drm/i915/intel_bios.h
@@ -239,6 +239,10 @@ struct common_child_dev_config {
 	u8 not_common2[2];
 	u8 ddc_pin;
 	u16 edid_ptr;
+	u8 obsolete;
+	u8 flags_1;
+	u8 not_common3[13];
+	u8 iboost_level;
 } __packed;
 
 /* This field changes depending on the BDB version, so the most reliable way to
-- 
2.1.4

